marlowe-playground-client/test/Marlowe/ContractTests.purs

```diff
@@ -20,91 +27,262 @@ import Examples.PureScript.EscrowWithCollateral as EscrowWithCollateral
+    equal (Just Close) (fromTerm =<< finalContract)
+    pure unit
+
+--
+exampleContractsDontHaveErrors :: TestSuite
+exampleContractsDontHaveErrors =
+  suite "Provided Examples don't throw errors nor have warnings" do
+    contractDontHaveErrors "Simple Escrow" filledEscrow
+    contractDontHaveErrors "Escrow with collateral" filledEscrowWithCollateral
+    contractDontHaveErrors "Zero coupon bond" filledZeroCouponBond
+    contractDontHaveErrors "Coupon bond guaranteed" filledCouponBondGuaranteed
+    contractDontHaveErrors "Swap" filledSwap
+    contractDontHaveErrors "Contract for differences" filledContractForDifferences
+    contractDontHaveErrors "Contract for differences with oracle" filledContractForDifferencesWithOracle
+
+-- This is a property based test that checks that for a given contract, the possible actions available
+-- during the simulation don't throw errors nor warnings.
+contractDontHaveErrors :: String -> Term T.Contract -> TestSuite
+contractDontHaveErrors contractName contract =
+  test contractName
+    $ quickCheck
+    $ evalStateT property mkState
+  where
+  property :: StateT Simulation.State Gen Result
+  property = do
+    startSimulation zero contract
+    runContract (Just contract)
+
+  -- When the possible action is a Choose, we need to generate a random value inside the provided Bound.
+  -- We use this function (which looses precision) in order to select a small integer and wrap it in a BigInteger
+  looselyToInt :: BigInteger -> Int
+  looselyToInt = round <<< BigInteger.toNumber
+
+  genValueInBound :: forall m. MonadGen m => MonadRec m => Bound -> m BigInteger
+  genValueInBound (Bound from to) = BigInteger.fromInt <$> chooseInt (looselyToInt from) (looselyToInt to)
+
+  runContract :: Maybe (Term T.Contract) -> StateT Simulation.State Gen Result
+  runContract Nothing = pure Success
+
+  runContract (Just (Term T.Close _)) = pure Success
+
+  -- TODO: For the moment it was not needed, as none of the examples triggered a warning nor an error
+  --       but we should add an extra parameter to runContract that includes the current Action list
+  --       and in case of a failure, we should print what is the list that causes the problem.
+  runContract (Just _) = do
+    simulationState <- get
+    let
+      mPossibleActions :: Maybe (NonEmptyArray ActionInput)
+      mPossibleActions =
+        simulationState
+          # preview _currentPossibleActions
+          # map getAllActions
+          # bindFlipped fromArray
+    case mPossibleActions of
+      Nothing -> pure $ Failed "no available actions"
+      Just possibleActions -> do
+        action <- lift $ elements possibleActions
+        case action of
+          MoveToSlot slot -> moveToSlot slot
+          DepositInput accountId party token value -> applyInput $ IDeposit accountId party token value
+          ChoiceInput choiceId bounds value -> do
+            randomValue <- lift $ oneOf $ pure value :| (genValueInBound <$> bounds)
```

Comment:
Just a thought: Usually values are generated outside the property. I wonder whether it would be beneficial, it does link with the `TODO` you wrote at the beginning of the function. But I see how it could complicate things, so maybe it is just a can of worms

Reply:
The TODO at the beginning of the function actually relates to the `Failed "It has errors"` and `Failed "it has warnings"` from below.
 
I thought of a different refactor to separate the property being tested from randomly recursing into a contract. To do this, we would need to extract the `case mError, mWarnings` into a function that runs on `StateT Simulation.State Gen Result` and can test whatever property it wants. But I'd do that if we find another property we want to uphold for all contracts. Maybe that the balances should always end in 0?

Reply:
Ah, I see, it is the Action list not the Input list that you were talking about. Yes, I guess reusing the traversal code for several properties is one advantage, makes sense